scrapy/http/response/text.py

```diff
@@ -158,6 +159,66 @@ def follow(self, url, callback=None, method='GET', headers=None, body=None,
             cb_kwargs=cb_kwargs,
         )
 
+    def follow_all(self, urls=None, callback=None, method='GET', headers=None, body=None,
+                   cookies=None, meta=None, encoding=None, priority=0,
+                   dont_filter=False, errback=None, cb_kwargs=None,
+                   css=None, xpath=None):
+        # type: (...) -> Generator[Request, None, None]
+        """
+        A generator that produces :class:`~.Request` instances to follow all
+        links in ``urls``. It accepts the same arguments as the :class:`~.Request`
+        initializer, except that each ``urls`` element does not need to be an absolute
+        URL, it can be any of the following:
+
+        * a relative URL;
+        * a scrapy.link.Link object (e.g. a link extractor result);
+        * an attribute Selector (not SelectorList) - e.g.
+          ``response.css('a::attr(href)')[0]`` or
+          ``response.xpath('//img/@src')[0]``.
+        * a Selector for ``<a>`` or ``<link>`` element, e.g.
+          ``response.css('a.my_link')[0]``.
+
+        In addition, ``css`` and ``xpath`` arguments are accepted to perform the link extraction
+        within the ``follow_all`` method (only one of ``urls``, ``css`` and ``xpath`` are accepted).
+
+        Note that when using the ``css`` or ``xpath`` parameters, this method will not produce
+        requests for selectors from which links cannot be obtained (for instance, anchor tags
+        without ``href`` attribute)
+        """
+        arg_count = len(list(filter(None, (urls, css, xpath))))
+        if arg_count != 1:
+            raise ValueError('Please supply exactly one of the following arguments: {urls, css, xpath}')
+        if not urls:
+            urls = []
+            if css:
+                selector_method = getattr(self, 'css')
+                expression = css
+            elif xpath:
+                selector_method = getattr(self, 'xpath')
+                expression = xpath
+            for selector in selector_method(expression):
+                try:
+                    urls.append(_url_from_selector(selector))
+                except ValueError:
+                    pass
+        return (
```

Comment:
My first implementation was like that, the problem is that the checks that raise exceptions are not executed until the result is first iterated. I'm all about generators and not putting everything in memory if it's not strictly necessary, but since the `css` and `xpath` methods return selector lists, there is not really a lot to be done in that regard.